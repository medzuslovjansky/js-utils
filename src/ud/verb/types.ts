import type { Word } from '../types';

export interface Aux extends Word {
  uPosTag: 'AUX';
}

export interface Verb extends Word {
  uPosTag: 'VERB';
  feats: VerbFeatures;
}

export interface VerbFeatures {
  aspect?: 'Imp' | 'Perf';
  gender?: 'Masc' | 'Fem' | 'Neut';
  mood?: 'Ind' | 'Cnd' | 'Imp' | 'Sub';
  number?: 'Sing' | 'Plur';
  person?: '1' | '2' | '3';
  tense?: 'Past' | 'Pres' | 'Fut' | 'Imp';
  verbForm?: 'Fin' | 'Inf' | 'Part' | 'Conv';
  voice?: 'Act' | 'Pass';
}
