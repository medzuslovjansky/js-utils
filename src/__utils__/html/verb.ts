/* eslint-disable prettier/prettier */
import type { SteenVerbParadigm } from '../../verb';
import * as common from './common';

export function render(value: SteenVerbParadigm) {
  return template(common.get(value));
}

export function renderDiff(a: SteenVerbParadigm, b: SteenVerbParadigm) {
  return template(common.diff(a, b));
}

function template(r: (key: keyof SteenVerbParadigm, index?: number) => string | null): string {
  const figure = `\
<figure>
   <figcaption>Verb conjugation</figcaption>
   <table>
      <thead>
         <tr>
            <th>person</th>
            <th>pronoun</th>
            <th>present</th>
            <th>imperfect</th>
            <th>future</th>
         </tr>
      </thead>
      <tbody>
         <tr>
            <th>1.sg</th>
            <td>ja</td>
            <td>${r('present', 0) || '—'}</td>
            <td>${r('imperfect', 0) || '—'}</td>
            <td>${r('future', 0) || '—'}</td>
         </tr>
         <tr>
            <th>2.sg</th>
            <td>ty</td>
            <td>${r('present', 1) || '—'}</td>
            <td>${r('imperfect', 1) || '—'}</td>
            <td>${r('future', 1) || '—'}</td>
         </tr>
         <tr>
            <th>3.sg</th>
            <td>on, ona, ono</td>
            <td>${r('present', 2) || '—'}</td>
            <td>${r('imperfect', 2) || '—'}</td>
            <td>${r('future', 2) || '—'}</td>
         </tr>
         <tr>
            <th>1.pl</th>
            <td>my</td>
            <td>${r('present', 3) || '—'}</td>
            <td>${r('imperfect', 3) || '—'}</td>
            <td>${r('future', 3) || '—'}</td>
         </tr>
         <tr>
            <th>2.pl</th>
            <td>vy</td>
            <td>${r('present', 4) || '—'}</td>
            <td>${r('imperfect', 4) || '—'}</td>
            <td>${r('future', 4) || '—'}</td>
         </tr>
         <tr>
            <th>3.pl</th>
            <td>oni, one</td>
            <td>${r('present', 5) || '—'}</td>
            <td>${r('imperfect', 5) || '—'}</td>
            <td>${r('future', 5) || '—'}</td>
         </tr>
      </tbody>
      </table>
      <table>
      <thead>
         <tr>
            <th>person</th>
            <th>pronoun</th>
            <th>perfect</th>
            <th>pluperfect</th>
            <th>conditional</th>
         </tr>
         <tr>
            <th>1.sg</th>
            <td>ja</td>
            <td>${r('perfect', 0) || '—'}</td>
            <td>${r('pluperfect', 0) || '—'}</td>
            <td>${r('conditional', 0) || '—'}</td>
         </tr>
         <tr>
            <th>2.sg</th>
            <td>ty</td>
            <td>${r('perfect', 1) || '—'}</td>
            <td>${r('pluperfect', 1) || '—'}</td>
            <td>${r('conditional', 1) || '—'}</td>
         </tr>
         <tr>
            <th rowspan="3">3.sg</th>
            <td>on</td>
            <td>${r('perfect', 2) || '—'}</td>
            <td>${r('pluperfect', 2) || '—'}</td>
            <td>${r('conditional', 2) || '—'}</td>
         </tr>
         <tr>
            <td>ona</td>
            <td>${r('perfect', 3) || '—'}</td>
            <td>${r('pluperfect', 3) || '—'}</td>
            <td>${r('conditional', 3) || '—'}</td>
         </tr>
         <tr>
            <td>ono</td>
            <td>${r('perfect', 4) || '—'}</td>
            <td>${r('pluperfect', 4) || '—'}</td>
            <td>${r('conditional', 4) || '—'}</td>
         </tr>
         <tr>
            <th>1.pl</th>
            <td>my</td>
            <td>${r('perfect', 5) || '—'}</td>
            <td>${r('pluperfect', 5) || '—'}</td>
            <td>${r('conditional', 5) || '—'}</td>
         </tr>
         <tr>
            <th>2.pl</th>
            <td>vy</td>
            <td>${r('perfect', 6) || '—'}</td>
            <td>${r('pluperfect', 6) || '—'}</td>
            <td>${r('conditional', 6) || '—'}</td>
         </tr>
         <tr>
            <th>3.pl</th>
            <td>oni, one</td>
            <td>${r('perfect', 7) || '—'}</td>
            <td>${r('pluperfect', 7) || '—'}</td>
            <td>${r('conditional', 7) || '—'}</td>
         </tr>
       </thead>
       </table>
       <table>
       <tbody>
         <tr>
            <th align="left" colspan="2">infinitive</th>
            <td colspan="3">${r('infinitive') || '—'}</td>
         </tr>
         <tr>
            <th align="left" colspan="2">imperative</th>
            <td colspan="3">${r('imperative') || '—'}</td>
         </tr>
         <tr>
            <th align="left" colspan="2">present active participle</th>
            <td colspan="3">${r('prap') || '—'}</td>
         </tr>
         <tr>
            <th align="left" colspan="2">present passive participle</th>
            <td colspan="3">${r('prpp') || '—'}</td>
         </tr>
         <tr>
            <th align="left" colspan="2">past active participle</th>
            <td colspan="3">${r('pfap') || '—'}</td>
         </tr>
         <tr>
            <th align="left" colspan="2">past passive participle</th>
            <td colspan="3">${r('pfpp') || '—'}</td>
         </tr>
         <tr>
            <th align="left" colspan="2">verbal noun</th>
            <td colspan="3">${r('gerund') || '—'}</td>
         </tr>
      </tbody>
   </table>
</figure>
`;

  return figure;
}
