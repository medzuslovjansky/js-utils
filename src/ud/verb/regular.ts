import { conjugationVerb, SteenVerbParadigm } from '../../verb';
import { Verb, VerbFeatures } from './types';

export function getRegularForms(
  lemma: string,
  additional: string,
  inheritedFeats: VerbFeatures,
): Verb[] {
  const paradigm = conjugationVerb(lemma, additional);
  if (!paradigm) {
    throw new Error(`No paradigm found for verb ${lemma}`);
  }

  return [
    getInfinitive,
    getPresentSg1,
    getPresentSg2,
    getPresentSg3,
    getPresentPl1,
    getPresentPl2,
    getPresentPl3,
    // ...
    getImperative,
  ].flatMap((callback): Verb[] => {
    const maybeForms = callback(paradigm);
    const forms = Array.isArray(maybeForms) ? maybeForms : [maybeForms];
    const result: Verb[] = [];
    for (const item of forms) {
      const { form, feats } = item;

      result.push({
        form,
        lemma: paradigm.infinitive,
        uPosTag: 'VERB',
        feats: {
          ...inheritedFeats,
          ...feats,
        },
      });
    }

    return result;
  });
}

type VerbForm = Pick<Verb, 'form' | 'feats'>;

function getPresentSg1(paradigm: SteenVerbParadigm): VerbForm {
  return {
    form: paradigm.present[0],
    feats: {
      mood: 'Ind',
      tense: 'Pres',
      person: '1',
      number: 'Sing',
    },
  };
}

function getPresentSg2(paradigm: SteenVerbParadigm): VerbForm {
  return {
    form: paradigm.present[1],
    feats: {
      mood: 'Ind',
      tense: 'Pres',
      person: '2',
      number: 'Sing',
    },
  };
}

function getPresentSg3(paradigm: SteenVerbParadigm): VerbForm {
  return {
    form: paradigm.present[2],
    feats: {
      mood: 'Ind',
      tense: 'Pres',
      person: '3',
      number: 'Sing',
    },
  };
}

function getPresentPl1(paradigm: SteenVerbParadigm): VerbForm {
  return {
    form: paradigm.present[3],
    feats: {
      mood: 'Ind',
      tense: 'Pres',
      person: '1',
      number: 'Plur',
    },
  };
}

function getPresentPl2(paradigm: SteenVerbParadigm): VerbForm {
  return {
    form: paradigm.present[4],
    feats: {
      mood: 'Ind',
      tense: 'Pres',
      person: '2',
      number: 'Plur',
    },
  };
}

function getPresentPl3(paradigm: SteenVerbParadigm): VerbForm {
  return {
    form: paradigm.present[5],
    feats: {
      mood: 'Ind',
      tense: 'Pres',
      person: '3',
      number: 'Plur',
    },
  };
}

// Add similar functions for the remaining verb forms (imperfect, perfect, pluperfect, future, conditional)

function getInfinitive(paradigm: SteenVerbParadigm): VerbForm {
  return {
    form: paradigm.infinitive,
    feats: {
      verbForm: 'Inf',
    },
  };
}

function getImperative(paradigm: SteenVerbParadigm): VerbForm[] {
  const [sg2, pl1, pl2] = paradigm.imperative.split(', ');
  const mood = 'Imp';

  return [
    {
      form: sg2,
      feats: { mood, person: '2', number: 'Sing' },
    },
    {
      form: pl1,
      feats: { mood, person: '1', number: 'Plur' },
    },
    {
      form: pl2,
      feats: { mood, person: '2', number: 'Plur' },
    },
  ];
}
