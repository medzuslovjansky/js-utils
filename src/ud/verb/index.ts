import { parsePos } from '../../partOfSpeech';
import type { Word } from '../types';
import { getRegularForms } from './regular';

export function forms(
  lemma: string,
  additional: string,
  partOfSpeech: string,
): Word[] {
  const pos = parsePos(partOfSpeech);
  if (pos.name !== 'verb') {
    throw new Error(`Expected a verb, but got: ${partOfSpeech}`);
  }

  let result: Word[] = [];
  // TODO: Add auxiliary forms
  if (!pos.auxiliary) {
    if (pos.perfective) {
      result = result.concat(
        getRegularForms(lemma, additional, { aspect: 'Perf' }),
      );
    }

    if (pos.imperfective) {
      result = result.concat(
        getRegularForms(lemma, additional, { aspect: 'Imp' }),
      );
    }
  }

  return result;
}
