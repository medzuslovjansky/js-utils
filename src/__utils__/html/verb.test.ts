/* eslint-disable @typescript-eslint/no-non-null-assertion */

import * as verb from './verb';
import { conjugationVerb } from '../../verb';

describe('renderers', () => {
  describe('HTML', () => {
    describe('verb conjugation', () => {
      test('render', () => {
        const dreti = conjugationVerb('drěti', '(dere)')!;
        expect(verb.render(dreti)).toMatchSnapshot();
      });

      test('renderDiff', () => {
        const before = conjugationVerb('dreti', '(dere)')!;
        // we choose a different verb to show the diff
        const after = conjugationVerb('trěti', '(tre)')!;
        expect(verb.renderDiff(before, after)).toMatchSnapshot();
      });
    });
  });
});
