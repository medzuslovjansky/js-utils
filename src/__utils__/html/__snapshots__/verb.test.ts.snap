// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderers HTML verb conjugation render 1`] = `
"<figure>
   <figcaption>Verb conjugation</figcaption>
   <table>
      <thead>
         <tr>
            <th>person</th>
            <th>pronoun</th>
            <th>present</th>
            <th>imperfect</th>
            <th>future</th>
         </tr>
      </thead>
      <tbody>
         <tr>
            <th>1.sg</th>
            <td>ja</td>
            <td>derų, derem</td>
            <td>drěh</td>
            <td>bųdų drěti</td>
         </tr>
         <tr>
            <th>2.sg</th>
            <td>ty</td>
            <td>dereš</td>
            <td>drěše</td>
            <td>bųdeš drěti</td>
         </tr>
         <tr>
            <th>3.sg</th>
            <td>on, ona, ono</td>
            <td>dere</td>
            <td>drěše</td>
            <td>bųde drěti</td>
         </tr>
         <tr>
            <th>1.pl</th>
            <td>my</td>
            <td>deremo</td>
            <td>drěhmo</td>
            <td>bųdemo drěti</td>
         </tr>
         <tr>
            <th>2.pl</th>
            <td>vy</td>
            <td>derete</td>
            <td>drěste</td>
            <td>bųdete drěti</td>
         </tr>
         <tr>
            <th>3.pl</th>
            <td>oni, one</td>
            <td>derųt</td>
            <td>drěhų</td>
            <td>bųdųt drěti</td>
         </tr>
      </tbody>
      </table>
      <table>
      <thead>
         <tr>
            <th>person</th>
            <th>pronoun</th>
            <th>perfect</th>
            <th>pluperfect</th>
            <th>conditional</th>
         </tr>
         <tr>
            <th>1.sg</th>
            <td>ja</td>
            <td>jesm dŕl(a)</td>
            <td>běh dŕl(a)</td>
            <td>byh dŕl(a)</td>
         </tr>
         <tr>
            <th>2.sg</th>
            <td>ty</td>
            <td>jesi dŕl(a)</td>
            <td>běše dŕl(a)</td>
            <td>bys dŕl(a)</td>
         </tr>
         <tr>
            <th rowspan="3">3.sg</th>
            <td>on</td>
            <td>(je) dŕl</td>
            <td>běše dŕl</td>
            <td>by dŕl</td>
         </tr>
         <tr>
            <td>ona</td>
            <td>(je) dŕla</td>
            <td>běše dŕla</td>
            <td>by dŕla</td>
         </tr>
         <tr>
            <td>ono</td>
            <td>(je) dŕlo</td>
            <td>běše dŕlo</td>
            <td>by dŕlo</td>
         </tr>
         <tr>
            <th>1.pl</th>
            <td>my</td>
            <td>jesmo dŕli</td>
            <td>běhmo dŕli</td>
            <td>byhmo dŕli</td>
         </tr>
         <tr>
            <th>2.pl</th>
            <td>vy</td>
            <td>jeste dŕli</td>
            <td>běste dŕli</td>
            <td>byste dŕli</td>
         </tr>
         <tr>
            <th>3.pl</th>
            <td>oni, one</td>
            <td>(sųt) dŕli</td>
            <td>běhų dŕli</td>
            <td>by dŕli</td>
         </tr>
       </thead>
       </table>
       <table>
       <tbody>
         <tr>
            <th align="left" colspan="2">infinitive</th>
            <td colspan="3">drěti</td>
         </tr>
         <tr>
            <th align="left" colspan="2">imperative</th>
            <td colspan="3">deri, derimo, derite</td>
         </tr>
         <tr>
            <th align="left" colspan="2">present active participle</th>
            <td colspan="3">derųći (derųća, derųće)</td>
         </tr>
         <tr>
            <th align="left" colspan="2">present passive participle</th>
            <td colspan="3">deremy (derema, deremo)</td>
         </tr>
         <tr>
            <th align="left" colspan="2">past active participle</th>
            <td colspan="3">dŕši (dŕša, dŕše)</td>
         </tr>
         <tr>
            <th align="left" colspan="2">past passive participle</th>
            <td colspan="3">dŕty (dŕta, dŕto)</td>
         </tr>
         <tr>
            <th align="left" colspan="2">verbal noun</th>
            <td colspan="3">dŕt́je</td>
         </tr>
      </tbody>
   </table>
</figure>
"
`;

exports[`renderers HTML verb conjugation renderDiff 1`] = `
"<figure>
   <figcaption>Verb conjugation</figcaption>
   <table>
      <thead>
         <tr>
            <th>person</th>
            <th>pronoun</th>
            <th>present</th>
            <th>imperfect</th>
            <th>future</th>
         </tr>
      </thead>
      <tbody>
         <tr>
            <th>1.sg</th>
            <td>ja</td>
            <td><del>derų, derem</del> <ins>trų, trem</ins></td>
            <td><del>dreh</del> <ins>trěh</ins></td>
            <td><del>bųdų dreti</del> <ins>bųdų trěti</ins></td>
         </tr>
         <tr>
            <th>2.sg</th>
            <td>ty</td>
            <td><del>dereš</del> <ins>treš</ins></td>
            <td><del>dreše</del> <ins>trěše</ins></td>
            <td><del>bųdeš dreti</del> <ins>bųdeš trěti</ins></td>
         </tr>
         <tr>
            <th>3.sg</th>
            <td>on, ona, ono</td>
            <td><del>dere</del> <ins>tre</ins></td>
            <td><del>dreše</del> <ins>trěše</ins></td>
            <td><del>bųde dreti</del> <ins>bųde trěti</ins></td>
         </tr>
         <tr>
            <th>1.pl</th>
            <td>my</td>
            <td><del>deremo</del> <ins>tremo</ins></td>
            <td><del>drehmo</del> <ins>trěhmo</ins></td>
            <td><del>bųdemo dreti</del> <ins>bųdemo trěti</ins></td>
         </tr>
         <tr>
            <th>2.pl</th>
            <td>vy</td>
            <td><del>derete</del> <ins>trete</ins></td>
            <td><del>dreste</del> <ins>trěste</ins></td>
            <td><del>bųdete dreti</del> <ins>bųdete trěti</ins></td>
         </tr>
         <tr>
            <th>3.pl</th>
            <td>oni, one</td>
            <td><del>derųt</del> <ins>trųt</ins></td>
            <td><del>drehų</del> <ins>trěhų</ins></td>
            <td><del>bųdųt dreti</del> <ins>bųdųt trěti</ins></td>
         </tr>
      </tbody>
      </table>
      <table>
      <thead>
         <tr>
            <th>person</th>
            <th>pronoun</th>
            <th>perfect</th>
            <th>pluperfect</th>
            <th>conditional</th>
         </tr>
         <tr>
            <th>1.sg</th>
            <td>ja</td>
            <td><del>jesm dŕl(a)</del> <ins>jesm tŕl(a)</ins></td>
            <td><del>běh dŕl(a)</del> <ins>běh tŕl(a)</ins></td>
            <td><del>byh dŕl(a)</del> <ins>byh tŕl(a)</ins></td>
         </tr>
         <tr>
            <th>2.sg</th>
            <td>ty</td>
            <td><del>jesi dŕl(a)</del> <ins>jesi tŕl(a)</ins></td>
            <td><del>běše dŕl(a)</del> <ins>běše tŕl(a)</ins></td>
            <td><del>bys dŕl(a)</del> <ins>bys tŕl(a)</ins></td>
         </tr>
         <tr>
            <th rowspan="3">3.sg</th>
            <td>on</td>
            <td><del>(je) dŕl</del> <ins>(je) tŕl</ins></td>
            <td><del>běše dŕl</del> <ins>běše tŕl</ins></td>
            <td><del>by dŕl</del> <ins>by tŕl</ins></td>
         </tr>
         <tr>
            <td>ona</td>
            <td><del>(je) dŕla</del> <ins>(je) tŕla</ins></td>
            <td><del>běše dŕla</del> <ins>běše tŕla</ins></td>
            <td><del>by dŕla</del> <ins>by tŕla</ins></td>
         </tr>
         <tr>
            <td>ono</td>
            <td><del>(je) dŕlo</del> <ins>(je) tŕlo</ins></td>
            <td><del>běše dŕlo</del> <ins>běše tŕlo</ins></td>
            <td><del>by dŕlo</del> <ins>by tŕlo</ins></td>
         </tr>
         <tr>
            <th>1.pl</th>
            <td>my</td>
            <td><del>jesmo dŕli</del> <ins>jesmo tŕli</ins></td>
            <td><del>běhmo dŕli</del> <ins>běhmo tŕli</ins></td>
            <td><del>byhmo dŕli</del> <ins>byhmo tŕli</ins></td>
         </tr>
         <tr>
            <th>2.pl</th>
            <td>vy</td>
            <td><del>jeste dŕli</del> <ins>jeste tŕli</ins></td>
            <td><del>běste dŕli</del> <ins>běste tŕli</ins></td>
            <td><del>byste dŕli</del> <ins>byste tŕli</ins></td>
         </tr>
         <tr>
            <th>3.pl</th>
            <td>oni, one</td>
            <td><del>(sųt) dŕli</del> <ins>(sųt) tŕli</ins></td>
            <td><del>běhų dŕli</del> <ins>běhų tŕli</ins></td>
            <td><del>by dŕli</del> <ins>by tŕli</ins></td>
         </tr>
       </thead>
       </table>
       <table>
       <tbody>
         <tr>
            <th align="left" colspan="2">infinitive</th>
            <td colspan="3"><del>dreti</del> <ins>trěti</ins></td>
         </tr>
         <tr>
            <th align="left" colspan="2">imperative</th>
            <td colspan="3"><del>deri, derimo, derite</del> <ins>tri, trimo, trite</ins></td>
         </tr>
         <tr>
            <th align="left" colspan="2">present active participle</th>
            <td colspan="3"><del>derųći (derųća, derųće)</del> <ins>trųći (trųća, trųće)</ins></td>
         </tr>
         <tr>
            <th align="left" colspan="2">present passive participle</th>
            <td colspan="3"><del>deremy (derema, deremo)</del> <ins>tremy (trema, tremo)</ins></td>
         </tr>
         <tr>
            <th align="left" colspan="2">past active participle</th>
            <td colspan="3"><del>dŕši (dŕša, dŕše)</del> <ins>tŕši (tŕša, tŕše)</ins></td>
         </tr>
         <tr>
            <th align="left" colspan="2">past passive participle</th>
            <td colspan="3"><del>dŕty (dŕta, dŕto)</del> <ins>tŕty (tŕta, tŕto)</ins></td>
         </tr>
         <tr>
            <th align="left" colspan="2">verbal noun</th>
            <td colspan="3"><del>dŕt́je</del> <ins>tŕt́je</ins></td>
         </tr>
      </tbody>
   </table>
</figure>
"
`;
